document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoUtils.java

```diff
@@ -145,6 +150,20 @@ public static ReturnDocument getReturnDocument(final ReturnDocumentType returnDo
                     String.format("Unhandled return document type: %s", returnDocumentType)));
   }
 
+  public static boolean isFunctionExpressionSelectionWithGroupBy(
+      final SelectionSpec selectionSpec, final List<String> groupByAliases) {
+    return selectionSpec.getAlias() != null
+        && groupByAliases.contains(selectionSpec.getAlias())
+        && selectionSpec.getExpression().getClass().equals(FunctionExpression.class);
+  }
+
+  public static List<String> getGroupByAliases(final List<GroupTypeExpression> expressions) {
+    return expressions.stream()
+        .filter(expression -> expression.getClass().equals(IdentifierExpression.class))
+        .map(expression -> ((IdentifierExpression) expression).getName())
```

Comment:
Could you avoid explicit instance of/class equality checks and casting by leveraging the visitor pattern?